tests/eve_tests/unit_tests/test_datamodels.py

```diff
@@ -0,0 +1,842 @@
+    ("typing.Optional[int]", [1, None], [[1], [], 1j]),
+    (
+        "typing.Dict[Union[int, float, str], Union[Tuple[int, Optional[float]], Set[int]]]",
+        [{1: (2, 3.0)}, {1.0: (2, None)}, {"1": {1, 2}}],
+        [{(1, 1.0, "1"): set()}, {1: [1]}, {"1": (1,)}],
+    ),
+]
+
+
+@pytest.mark.parametrize(["type_hint", "valid_values", "wrong_values"], sample_type_data)
+def test_field_type_hint(type_hint: str, valid_values: Sequence[Any], wrong_values: Sequence[Any]):
+    context: Dict[str, Any] = {}
+    exec(
+        f"""
+@datamodels.datamodel
+class Model:
+    value: {type_hint}
+""",
+        globals(),
+        context,
+    )
+    Model = context["Model"]
+
+    field_type = Model.__datamodel_fields__.value.type
+    if isinstance(field_type, str):
+        field_type = eval(field_type)
+    assert field_type == eval(type_hint)
+
+    for value in valid_values:
+        Model(value=value)
+
+    for value in wrong_values:
+        with pytest.raises((TypeError, ValueError), match="'value'"):
+            Model(value=value)
+
+
+def test_custom_class_type_hint():
+    @datamodels.datamodel
+    class Model1:
+        value: SampleEnum
+
+    Model1(value=SampleEnum.FOO)
+
+    class SampleClassChild(SampleClass):
+        pass
+
+    class SampleClassGrandChild(SampleClassChild):
+        pass
+
+    @datamodels.datamodel
+    class Model2:
+        value: SampleClass
+
+    Model2(value=SampleClass())
+    Model2(value=SampleClassChild())
+    Model2(value=SampleClassGrandChild())
+
+    class AnotherClass:
+        pass
+
+    with pytest.raises(TypeError, match="value"):
+        Model1(value=AnotherClass())
+    with pytest.raises(TypeError, match="value"):
+        Model2(value=AnotherClass())
+
+
+class MyType:
+    def __init__(self, value: Any) -> None:
+        self.value = value
+
+    def add(self, something) -> None:
+        return self.value + something
+
+    @classmethod
+    def __type_validator__(cls) -> datamodels.ValidatorType:
+        def _custom_validator(
+            instance: datamodels.DataModelTp, attribute: datamodels.Attribute, value: Any
+        ) -> None:
+            if not (hasattr(value, "value") and hasattr(value, "add")):
+                raise TypeError("Invalid value type for '{attribute.name}' field.")
+
+        return _custom_validator
+
+
+def test_custom_type_hint_validator():
+    @datamodels.datamodel
+    class Model:
+        value: MyType
+
+    Model(value=types.SimpleNamespace(value=32, add=22))
+
+    with pytest.raises(TypeError, match="value"):
+        Model(value=types.SimpleNamespace(value=32))
+
+
+@datamodels.datamodel
+class GlobalRecursiveModel:
+    value: Optional[GlobalRecursiveModel] = None
+
+
+def test_deferred_class_type_hint():
+    # Model defined in a module global context
+    assert isinstance(GlobalRecursiveModel.__datamodel_fields__.value.type, ForwardRef)
+
+    m1 = GlobalRecursiveModel()
+    m2 = GlobalRecursiveModel(value=m1)
```

Comment:
Maybe a test for a model with a `Union` or `List` of models as a field would be useful to test the recursion of the type validator.

Reply:
Done

Reply:
By the way, `Optional[T]` is already a Union (`Union[T, NoneType]`)  :)